add typing indicator

gtalha07 opened the pull request:
### What does it do?

This PR implements the typing indicator which is missing from the package. The current change allow Chat widget to add optional typing indicator. It takes `List<types.User>` as a parameter which by default is empty. When empty, it also doesn't show any indicator. When we have more than 2 users in list, it shows small count bubble, otherwise it'll show two users avatars actively typing and indicator. The indicator widget has support of customization in the form of `bubbleBorder`, `bubbleColor`, `animatedCircleColor`,
 `ForwardAnimationDuration` and `ReverseAnimationDuration` to control animation duration of indicator circles. A lot of has been commented as per your pattern in the code. 

### Why is it needed?

This is most useful for people who want to show typing indicator in their chat implementations. I believe it'll help to make package even better. 

### How to test it?

From the example code. We can add users list as a parameter for `showTypingIndicator()`. By default its disabled and takes empty list.

### Related issues/PRs

This PR relates to #269.

demchenkoalex commented:
Thanks a lot! Will check this out ASAP.

gtalha07 commented:
![typingIndicatorDefault](https://user-images.githubusercontent.com/68579938/180438680-994ed391-1a51-47a6-bee2-ed798889f507.gif)
![typingIndicatorDark](https://user-images.githubusercontent.com/68579938/180439526-df4db087-7d07-4bf0-b166-30f2b4de2bb3.gif)

gtalha07 commented:
![Screenshot_1658493904](https://user-images.githubusercontent.com/68579938/180442145-42233c20-1e69-448e-aec5-05cccf9ef352.png)
![Screenshot_1658493938](https://user-images.githubusercontent.com/68579938/180442234-a18d5ba5-d486-4f3f-adb5-c2635cdfe4e2.png)

demchenkoalex commented:
Hey! Thanks again for the big contribution! I hope you don't mind a couple of requests:

1. I asked @dariakhimych our designer to ASAP prepare the design for this feature, I don't think that the typing indicator itself will change much, but after our brief conversation we were on the same page about the multiple users typing thing, that it will be text.
2. To make this library more contribution friendly, together with @felixgabler we decided to minimise amount of props we pass to the `Chat` itself by using 2 things: `Theme` and `Options`. Now `Theme` will be similar to the Flutter's `ThemeData` - you can find tons of other components themes inside `ThemeData`, like `CheckboxThemeData`, `DrawerThemeData`, `SnackBarThemeData` etc. In our case, `ChatTheme` should include `TypingIndicatorTheme`, that will contain all theme customisation for the typing indicator only like colours, borders etc. We don't have an example yet, but I think you can create `TypingIndicatorTheme` class in the same file as the component itself, and make it similar to the `InputOptions` class you can find in the `input.dart`. The other thing, `Options` (might not be applicable to your case) is anything else that is not theme, and not so important to have on the `Chat` itself. See `InputOptions` mentioned above for the example. `showTypingIndicator ` should be on the `Chat` cause it will be widely used, so that is OK.

felixgabler commented on lib/src/widgets/typing_indicator.dart:
```diff
@@ -0,0 +1,284 @@
+    required this.bubbleAlignment,
+  });
+
+  /// See [types.User].
```
This comment should actually explain what the parameter is for and not it's type

felixgabler reviewed (commented):
Looks great!

gtalha07 commented:
Added ```TypingIndicatorTheme``` in ```ChatTheme``` which will allow us to set custom theme for typing indicator. Removed count bubble for multiple users text as per according to your changes. Will look forward for that feature design. Animations for dots is set to ```800ms``` as it doesn't really comply with theme itself and is not important.

demchenkoalex commented:
[Flyer Chat - Typing.pdf](https://github.com/flyerhq/flutter_chat_ui/files/9193458/Flyer.Chat.-.Typing.pdf)

this is what @dariakhimych prepared. Neutral 1 color is - you can add it to the `chat_theme.dart`. Also feel free to not localise anything, I will add it myself with all the languages. Also pdf includes a link to the example animation our designer is expecting, similar to what you have, maybe different speeds/easings/sequence. 

```dart
const neutral1 = Color(0xff615e6e);
```

gtalha07 commented:
> [Flyer Chat - Typing.pdf](https://github.com/flyerhq/flutter_chat_ui/files/9193458/Flyer.Chat.-.Typing.pdf)
> 
> this is what @dariakhimych prepared. Neutral 1 color is - you can add it to the `chat_theme.dart`
> 
> ```dart
> const neutral1 = Color(0xff615e6e);
> ```
> 
> Also feel free to not localise anything, I will add it myself with all the languages. Also pdf includes a link to the example animation our designer is expecting, similar to what you have, maybe different speeds/easings/sequence.

One thing I want to add, think of it as suggestion. Since the typing indicator takes textStyle for multiple users. Why not we allow it to take as widget. ```Text``` is also a widget but introducing that will allow developers to put any widget inside it. For instance, it'll also allow us to put ```UserAvatar``` instead of text and more accessibility for those who needed to place avatars?

demchenkoalex commented:
Hey @gtalha07 what will you say if our designer will make an alternative solution with avatars, cause I don't think it will be easy to just give widget and lay out both text or avatars, since text just says "N others" and avatar like you suggested will have +4 or something. You think this is feasible? We will do like a mode "Text" / "Avatar"?

gtalha07 commented:
> Hey @gtalha07 what will you say if our designer will make an alternative solution with avatars, cause I don't think it will be easy to just give widget and lay out both text or avatars, since text just says "N others" and avatar like you suggested will have +4 or something. You think this is feasible? We will do like a mode "Text" / "Avatar"?

Hi @demchenkoalex, I think the possibility is there to make typing indicator toggle between text and avatars. This is just a demo of what we can do for avatars. 
![Screenshot_1659008720](https://user-images.githubusercontent.com/68579938/181497715-7f13bb82-c5b6-4af0-9e69-c1993cb997bc.png)

YassineChe commented:
Hello.
Can you please be more specific about this feature implementation, cuz i couldn't find `showTypingIndicator()` attribute.
Or please, show us a snippet,
Thanks!

demchenkoalex commented:
@YassineChe this is a pull request, where we discuss and finalise code before merging. Then, after a release, you will be able to use it.

demchenkoalex commented:
Hey @gtalha07, sorry for the delay, here is what our designer made. I can see here 3 modes actually: only text, only avatar or both.

![typing-with-avatars](https://user-images.githubusercontent.com/14123304/182436321-22c4ae1a-93bb-4f40-8a78-643c69b4a437.png)

gtalha07 commented:
@demchenkoalex I'm also looking here the support of ```indicatorBuilder``` which can allow to set up custom typing indicator widget. What do you think?

demchenkoalex commented:
Yes, @gtalha07 agree!

gtalha07 commented:
> Yes, @gtalha07 agree!

@demchenkoalex Please review it. I think its looking good now and cool package needs update. 😄

felixgabler commented on doc/advanced-usage.md:
```diff
@@ -555,3 +555,8 @@ class _MyHomePageState extends State<MyHomePage> {
 To show user avatars & names use `showUserAvatars` and `showUserNames` parameters. Can be used separately. By default, the chat will select one of 10 provided colors as an avatar background and name text color. Color is calculated based on the user's `id` hash code, so it is unique in different rooms. To modify provided colors use `userAvatarNameColors` parameter in [theme](themes). If you want t ...
+
+## Typing Indicator
+To show typing indicator, you can use `showTypingIndicator` optional parameter which takes [User.types] list. By default its empty list which hides the indicator. Shows typing user avatar along with animated speech bubble. You can set your own theme for typing indicator, see [theme](themes).
```
```suggestion
To show a typing indicator, you can use the optional parameter `showTypingIndicator` which requires a type of `List<types.User>`. If provided a non-empty list, it will show a typing user avatar along with animated speech bubble. If the list is empty, it will hide the indicator. To set your own theme for the typing indicator, see [theme](themes).
```

What exactly does theme resolve to?

demchenkoalex replied:
sections on the left side

<img width="312" alt="Screenshot 2022-08-06 at 13 10 00" src="https://user-images.githubusercontent.com/14123304/183246391-298da2b7-bbf1-4e4e-9d86-422f884a42b7.png">

but docs do not need a review, I will review them myself and re-write if needed.

felixgabler commented on lib/src/chat_theme.dart:
```diff
@@ -403,6 +412,19 @@ class DefaultChatTheme extends ChatTheme {
       height: 1.333,
     ),
+    super.typingIndicatorTheme = const TypingIndicatorTheme(
+      bubbleColor: neutral7,
```
sort properties alphabetically

felixgabler commented on lib/src/widgets/chat.dart:
```diff
@@ -3,25 +3,28 @@ import 'dart:math';
 import 'package:flutter/material.dart';
 import 'package:flutter_chat_types/flutter_chat_types.dart' as types;
 import 'package:intl/intl.dart';
-import 'package:photo_view/photo_view.dart' show PhotoViewComputedScale;
```
You're fixing conflicts commit [3bae598](https://github.com/flyerhq/flutter_chat_ui/pull/293/commits/3bae59834ff96db88158befdc99c029b281a4ff3) actually reverted some progress (i.e. extraction of `ImageGallery` widget and using `VoidCallback`. Please make sure to bring them back.

felixgabler commented on lib/src/widgets/typing_indicator.dart:
```diff
@@ -0,0 +1,440 @@
+
+  @override
+  Widget build(BuildContext context) {
+    size = MediaQuery.of(context).size;
```
I'd say its not good practice to retain build time constants like this outside of the build function. Better to pass them through where they are needed.

felixgabler commented on lib/src/widgets/typing_indicator.dart:
```diff
@@ -0,0 +1,440 @@
+                    child: _typingContainer(widget.indicatorMode),
+                  ),
+                )
+              : Container(),
```
empty `Container()` can be `const SizedBox()` to stay efficient. Our analysis warnings should report this already. Do you have them enabled?

gtalha07 replied:
I already had analysis checks enabled in my environment yet it didn't announced here. Will change it.

felixgabler commented on lib/src/widgets/typing_indicator.dart:
```diff
@@ -0,0 +1,440 @@
+
+  void _showIndicator() {
+    _appearanceController
+      ..duration = const Duration(milliseconds: 500)
```
It would be easier to set the duration during instantiation.

felixgabler commented on lib/src/widgets/typing_indicator.dart:
```diff
@@ -0,0 +1,440 @@
+
+  /// Display mode of typing.
+  Widget _typingContainer(TypingIndicatorMode mode) {
+    if (mode == TypingIndicatorMode.text) {
```
A `switch` would be nicer to get IDE warnings when the enum changes.

felixgabler commented on lib/src/widgets/typing_indicator.dart:
```diff
@@ -0,0 +1,440 @@
+    if (author.isEmpty) {
+      return '';
+    } else if (author.length == 1) {
+      return '${author.first.firstName} is typing';
```
internationalization required

gtalha07 replied:
This was discussed earlier and will be handled by @demchenkoalex himself.

demchenkoalex replied:
Yep, this is ok! Sorry :)

felixgabler commented on lib/src/widgets/typing_indicator.dart:
```diff
@@ -0,0 +1,440 @@
+  }
+
+  // Typing avatar Widget.
+  Widget _typingAvatar(types.User author) {
```
You could extract a widget here. The more the better :)

felixgabler commented on lib/src/widgets/typing_indicator.dart:
```diff
@@ -0,0 +1,440 @@
+  State<TypingIndicator> createState() => _TypingIndicatorState();
+}
+
+class _TypingIndicatorState extends State<TypingIndicator>
```
This class is very long. Please take a look whether you can extract more widgets. That would also be better for performance.

felixgabler commented on lib/src/widgets/typing_indicator.dart:
```diff
@@ -0,0 +1,440 @@
+  /// Handler for multi user typing avatars.
+  Widget _multiUserAvatarBuilder(List<types.User> author) {
+    if (author.isEmpty) {
+      return Container();
```
```suggestion
      return const SizedBox();
```

felixgabler commented on lib/src/widgets/typing_indicator.dart:
```diff
@@ -0,0 +1,440 @@
+}
+
+@immutable
+class TypingIndicatorTheme {
```
@demchenkoalex where do we want to keep the themes? With the widget itself or in the main theme file? For the unseen messages, I placed it in the theme file but I also like this approach.

demchenkoalex replied:
I thought with the widget, similar to options. Nice to have everything you need in one file.

felixgabler commented on lib/src/widgets/chat.dart:
```diff
@@ -81,10 +84,12 @@ class Chat extends StatefulWidget {
     this.scrollToUnseenOptions = const ScrollToUnseenOptions(),
     this.showUserAvatars = false,
     this.showUserNames = false,
+    this.showTypingIndicator = const [],
```
These options could also be added to an option class similar to `InputOptions`

gtalha07 commented on lib/src/widgets/typing_indicator.dart:
```diff
@@ -0,0 +1,440 @@
+      ..forward();
+  }
+
+  Animation<Offset> _firstCircleOffset() => TweenSequence<Offset>(
```
@felixgabler I'm trying to convert this method into reusable one like this
```_circleOffset(Offset? start, Offset? end, double startInterval, double endInterval)```
However we would need to define these as ```const``` before ```build```. Do you have any opinion whether this would be feasible or not?

gtalha07 replied:
Because we need to pass values to ```_animatedCirclesController``` in ```initState()```, if we go with that.

felixgabler replied:
First of all, good idea to make it reusable by parameterisation. It should lead to exactly the same behavior so I don't quite understand what you mean with const? The parameters for it would still remain `const`. Could you maybe send me some code that you think would break here?

gtalha07 replied:
Sorry not ```const``` but literal values like this:

```
void initState() {
    super.initState();

    //

    _firstCircleOffsetAnimation = _circleOffset(
      Offset.zero,
      const Offset(0.0, -0.90),
      0.0,
      1.0,
    );
    _secondCircleOffsetAnimation = _circleOffset(
      Offset.zero,
      const Offset(0.0, -0.80),
      0.30,
      1.0,
    );
    _thirdCircleOffsetAnimation = _circleOffset(
      Offset.zero,
      const Offset(0.0, -0.90),
      0.45,
      1.0,
    );
    
    //

  }
  ```
  **The method will be:**
  ```
  Animation<Offset> _circleOffset(
    Offset? start,
    Offset? end,
    double startInterval,
    double endInterval,
  ) =>
      TweenSequence<Offset>(
        <TweenSequenceItem<Offset>>[
          TweenSequenceItem<Offset>(
            tween: Tween<Offset>(
              begin: start,
              end: end,
            ),
            weight: 50.0,
          ),
          TweenSequenceItem<Offset>(
            tween: Tween<Offset>(
              begin: end,
              end: start,
            ),
            weight: 50.0,
          ),
        ],
      ).animate(CurvedAnimation(
        parent: _animatedCirclesController,
        curve: Interval(startInterval, endInterval, curve: Curves.linear),
        reverseCurve:
            Interval(startInterval, endInterval, curve: Curves.linear),
      ));
      
   ```

felixgabler replied:
You could also extract the `const Interval(startInterval, endInterval)` since it's used twice. Overall, extracting the method is definitely a code improvement so I'd say go for it!

gtalha07 commented:
@felixgabler Have a look at new code. Many changes have been done. It looks nice thanks to you!

- [ ] Extracted widgets where possible.
- [ ] Refactor methods into reusable one.
- [ ] Added ```typingIndicatorOptions```. This now takes author(s) list by same mechanisam. Removed ```showTypingIndicator```.
- [ ] Proposed changes as per review.

felixgabler commented on lib/src/widgets/typing_indicator.dart:
```diff
@@ -0,0 +1,498 @@
+}
+
+/// Animation Interval for circles.
+Interval _animationInterval(
```
I don't think this is really necessary because the two calls from within `_circleOffset` have the same arguments. I'd be in favour of making the third parameter for `_circleOffset` an interval directly. Then, it can also be const.

felixgabler commented on lib/src/widgets/typing_indicator.dart:
```diff
@@ -0,0 +1,498 @@
+
+  /// Author(s) for [TypingIndicator].
+  /// By default its empty list which hides the indicator, see [types.User].
+  final List<types.User> authors;
```
I'm wondering if `authors` really is a good name for this property. I think `typingUsers` would explain the function and type better, what do you think?

felixgabler commented on lib/src/widgets/typing_indicator.dart:
```diff
@@ -0,0 +1,498 @@
+  late Animation<Offset> _firstCircleOffsetAnimation;
+  late Animation<Offset> _secondCircleOffsetAnimation;
+  late Animation<Offset> _thirdCircleOffsetAnimation;
+  @override
```
nit: In general, I'd prefer to leave a space before messages for readability.

felixgabler commented on lib/src/widgets/typing_indicator.dart:
```diff
@@ -0,0 +1,498 @@
+
+  final Color circlesColor;
+  final Animation<Offset> animationOffset;
+  @override
```
nit: new line before method

felixgabler commented on lib/src/widgets/typing_indicator.dart:
```diff
@@ -0,0 +1,498 @@
+}
+
+/// Handler for multi user typing text.
+String _multiUserTextBuilder(List<types.User> author) {
```
This method should live in the class where it is used (i.e. `TypingWidget`)

felixgabler commented on lib/src/widgets/typing_indicator.dart:
```diff
@@ -0,0 +1,498 @@
+}
+
+/// Used to specify width of stacking avatars based on number of authors.
+double _getStackingWidth(List<types.User> author, double indicatorWidth) {
```
Same here.

felixgabler reviewed (commented):
Good work! We're almost there :)

@demchenkoalex only thing I am not sure about (aside from tiny nits) is whether we should move the extracted widgets into their own files. What do you think?

gtalha07 commented:
@felixgabler However with some new changes in [commit](https://github.com/flyerhq/flutter_chat_ui/commit/36e1dd44b735b65c1d8491198a4fcd3d1a31e58c), the typing indicator area seems to be fixed on screen like ```input``` widget. It doesn't scroll along with messages. Do I need scroll controller here?  You can test through ```example```.

felixgabler commented:
@gtalha07 I checked out your PR at different commits and this behaviour was already there before the commit you mentioned. To get the desired behavior, the typing indicator would need to be a sliver in the `CustomScrollView` in `chat_list.dart`. Is this something we want in the design?

gtalha07 commented:
> @gtalha07 I checked out your PR at different commits and this behaviour was already there before the commit you mentioned. To get the desired behavior, the typing indicator would need to be a sliver in the `CustomScrollView` in `chat_list.dart`. Is this something we want in the design?

It seems like it got changed upon adding code. @demchenkoalex didn't mentioned specifically anything about this but it's expected the widget should be before latest message.

felixgabler commented:
Then it will have to be moved somewhere else. It can't scroll with the messages if it's not part of the chat list. Right now, it's directly next to the Input and as such also behaves like that, i.e. is fixed

demchenkoalex commented:
First of all, thanks @gtalha07 so much for your work and thanks @felixgabler for reviewing this.

Regarding the position of the indicator among other widgets - on videos attached before, indicator correctly appeared and animated all messages to accommodate space for itself. We could observe a similar behaviour in Signal or Messenger. However, if the user scrolls chat even for like 10 pixels (i.e. we are not in the beginning of the scroll list), indicator will not push messages up, so it would behave like it never appeared, but then if you scroll back to the start of the list, it will be visible. So there are 2 cases:

1. List is not scrolled. Indicator appears and animates messages up.
2. List is scrolled for at least 1 pixel, indicator appears in the list, but does not scroll content up, user needs to scroll down to see it.

Does this makes sense? Will it be hard to do?

Another thing I noticed while looking at the videos above, that when indicator is removed there is no animation, it just disappears - will it be possible to keep it alive for a duration of the delete animation and then remove the instance?

gtalha07 commented:
The intentional animation is as follows:
![TIAnimation](https://user-images.githubusercontent.com/68579938/184047586-bb8fec72-41e1-4ee4-ac4d-4c527cb58eed.gif)
However I can't seem to call ```_appearanceController.reverse()``` before typing indicator ```dispose()``` method. Because on dispose, it just removes widget without animation even if we call reverse animation there. Any ideas of how to achieve this ? @demchenkoalex @felixgabler.

felixgabler commented:
> The intentional animation is as follows:
> ![TIAnimation](https://user-images.githubusercontent.com/68579938/184047586-bb8fec72-41e1-4ee4-ac4d-4c527cb58eed.gif)
> However I can't seem to call ```_appearanceController.reverse()``` before typing indicator ```dispose()``` method. Because on dispose, it just removes widget without animation even if we call reverse animation there. Any ideas of how to achieve this ? @demchenkoalex @felixgabler. 
> 

Either you make sure it does not get disposed until the animation finishes or we keep the typing indicator always alive. The former would be more efficient but probably have worse code because we need to either have callbacks or set two booleans (typing indicator showing and removed) when the typingUser array changes that control it's visibility with removed being delayed until animation finishes.

I'd be okay with always showing it but maybe add a feature flag then for people who are never going to use the typing indicator anyways. Then, they can prevent the minor overhead.

felixgabler commented on lib/src/widgets/chat_list.dart:
```diff
@@ -131,6 +143,25 @@ class _ChatListState extends State<ChatList>
           physics: widget.scrollPhysics,
           reverse: true,
           slivers: [
+            widget.typingIndicatorOptions!.customTypingIndicator != null
```
The `SliverPadding` and `SliverToBoxAdapter` are the same in both branches so it would be best to pull the ternary operator into there (at the child of `SliverToBoxAdapter`).

Also, you could consolidate the ternary operator using the null-coalescing operator.

```dart
child: widget.typingIndicatorOptions?.customTypingIndicator ?? TypingIndicator(..),
```

gtalha07 commented:
@demchenkoalex Are you referring to this behavior on scrolling content and typing indicator status?  
https://user-images.githubusercontent.com/68579938/185659347-e406cc4f-b608-49b0-960c-708db9876f0b.mp4

demchenkoalex commented:
will resolve this ASAP. I want to merge this and if I will see that something is not right I will try to fix it myself. Sorry for the waiting and thanks for the huge amount of work included here!

elhe26 commented:
> will resolve this ASAP. I want to merge this and if I will see that something is not right I will try to fix it myself. Sorry for the waiting and thanks for the huge amount of work included here!

Thanks, @demchenkoalex .  Is there any roadmap to follow?

demchenkoalex commented:
Hey @elhe26, the roadmap is straightforward - finish swipe to reply and then all highly requested features - reactions, audio, video messages, time inside bubbles. Everything else is a bonus :)

ramy9109 commented:
any update coming soon for typing Indicator?

demchenkoalex commented:
yes, I cleaned up tons of issues and MRs last week, so I can do this and replies now.

demchenkoalex commented:
Thanks for the PR!

demchenkoalex commented:
Just wanted to add @gtalha07 after all discussions this works perfectly how I imagined it to work, awesome job and sorry for the delay.

rlee1990 commented:
Any roadmap on when this will be live?

baptistesx commented:
@demchenkoalex any news on the availability of this nice feature? Thank you guys for your awesome work!

demchenkoalex commented:
why roadmap? it is already merged if you use the lib from GitHub directly, and to release on pub.dev I need to fix RTL languages, didn't find time yet.

baptistesx commented:
I see, my bad i wasn't understanding as i saw the merge but the feature wasn't available on pub, thanks!

baptistesx commented:
I got a quick question: how can we trigger the user is typing (to emit the event on the chat socket)? Any way to access the textfield onChange on something similar?

prkhrv commented:
Is this feature available on pub?

iqfareez commented:
@prkhrv I think not yet. But hopefully they can publish to pub asap

thomers commented:
Hello, I'd like to show a typing indicator like this:

<img width="225" alt="Screenshot 2024-06-29 at 11 10 48" src="https://github.com/flyerhq/flutter_chat_ui/assets/453568/cc009621-3c65-49a3-a3f3-fc8f1704a37b">

Is there documentation on this? What do I have to do? Thanks! Thomas

demchenkoalex commented:
start with `typingIndicatorOptions` param of the `Chat` and pass `TypingIndicatorOptions` class there. Pass users that are typing to the `typingUsers` array and the loading indicator will appear. Play with other params in `TypingIndicatorOptions` to achieve required style.

thomers commented:
I'm trying to show the avatar left from the typing indicator - what I get is this:

![IMG_C20C9028BB0F-1](https://github.com/flyerhq/flutter_chat_ui/assets/453568/2bc4231d-9c6f-4b1f-b0c5-490f61665f89)

The white background is somehow too small for the animation. 

Here's my code.  

` theme: const DefaultChatTheme(
                  seenIcon: Text('read', style: TextStyle(fontSize: 12.0)),
                  inputBackgroundColor: Colors.white12,
                  backgroundColor: Colors.transparent,
                  inputTextColor: Colors.white,
                  primaryColor: chatPrimaryColor,
                  secondaryColor: chatSecondaryColor,
                ),
                typingIndicatorOptions: TypingIndicatorOptions(
                  typingMode: TypingIndicatorMode.avatar,
                )`


How can I fix the background and show the avatar left from the animation? 

Thomas

PS: Thanks for this awesome package! Really helps a lot.

demchenkoalex commented:
not sure showing avatar left from animation is possible; If you have multiple people typing your animation might go out of the screen. White background seems like a bug, it should not have any background right?
